include/swift/Basic/ExperimentalDependencies.h

```diff
@@ -0,0 +1,968 @@
+public:
+  // Define this here so it can be changed easily.
+  template <typename Key, typename MapValue>
+  using Map = std::unordered_map<Key, MapValue>;
+
+  using InnerMap = Map<Key2, Value>;
+
+private:
+  Map<Key1, InnerMap> map;
+
+public:
+  Optional<Value> find(const Key1 &k1, const Key2 &k2) const {
+    auto iter = map.find(k1);
+    if (iter == map.end())
+      return None;
+    auto iter2 = iter->second.find(k2);
+    return iter2 == iter->second.end() ? None : Optional<Value>(iter2->second);
+  }
+
+  /// The sought value must be present.
+  Value findAndErase(const Key1 &k1, const Key2 &k2) {
+    auto &submap = map[k1];
+    auto iter = submap.find(k2);
+    assert(iter != submap.end() && "Cannot erase nonexistant element.");
+    Value v = iter->second;
+    submap.erase(iter);
+    return v;
+  }
+
+  bool insert(const Key1 &k1, const Key2 &k2, Value &v) {
+    return map[k1].insert(std::make_pair(k2, v)).second;
+  }
+
+  /// Move value from (old1, old2) to (new1, new2)
+  bool move(const Key1 &old1, const Key2 &old2, const Key1 &new1,
+            const Key2 &new2) {
+    Value v = findAndErase(old1, old2);
+    insert(old1, old2, v);
+  }
+
+  /// Returns the submap at \p k1. May create one if not present.
+  Map<Key2, Value> &operator[](const Key1 &k1) { return map[k1]; }
+
+  /// Invoke \p fn on each Key2 and Value matching (k, *)
+  void forEachValueMatching(
+      const Key1 &k1,
+      function_ref<void(const Key2 &, const Value &)> fn) const {
+    const auto &iter = map.find(k1);
+    if (iter == map.end())
+      return;
+    for (auto &p : iter->second)
+      fn(p.first, p.second);
+  }
+
+  /// Invoke \p fn for each entry
+  void forEachEntry(
+      function_ref<void(const Key1 &, const Key2 &, const Value &)> fn) const {
+    for (const auto &p : map)
+      for (const auto &p2 : p.second)
+        fn(p.first, p2.first, p2.second);
+  }
+
+  /// Invoke fn for each Key1 and submap
+  void
+  forEachKey1(function_ref<void(const Key1 &, const InnerMap &)> fn) const {
+    for (const auto &p : map)
+      fn(p.first, p.second);
+  }
+
+  /// Check integrity and call \p verifyFn for each element, so that element can
+  /// be verified.
+  /// Verify absence of duplicates and call \p verifyFn for each element.
+  void verify(function_ref<void(const Key1 &k1, const Key2 &k2, Value v)>
+                  verifyFn) const {
+    std::unordered_set<Value> vals;
+    for (const auto &p1 : map)
+      for (const auto &p2 : p1.second) {
+        const Key1 &k1 = p1.first;
+        const Key2 &k2 = p2.first;
+        Value const v = p2.second;
+        if (!vals.insert(v).second) {
+          llvm::errs() << "Duplicate value at (" << k1 << ", " << k2 << ") "
+                       << v << "\n";
+          llvm_unreachable("duplicate node");
+        }
+        verifyFn(k1, k2, v);
+      }
+  }
+};
+
+/// Double-indexing in either order; symmetric about key order.
+/// The DriverGraph needs this structure.
```

Comment:
Oof, expensive.

Reply:
Not sure what you are thinking...memory? I *think* that the either-order indexing allows for O(1) access for all needed operations at the cost of twice as many lookups for some things.

Reply:
Yeah, memory, plus making insertions more expensive. It's not even a "don't do this"; I just wrote down the thought as I had it. Maybe it'll turn out we don't actually need both mappings in the DriverGraph, but maybe we do and this really is the best structure.

Reply:
Thanks. I had reasoned it out and decided both were needed, but should double-check.

Reply:
Double-checking... I can see more than one place where each key is used first.